File: csv-query/csharp/src/CsvQuery/Query.cs
using System.Globalization;

namespace CsvQuery;

public class Query
{
    private readonly CsvTable _table;
    private List<Row> _rows;

    public Query(CsvTable table)
    {
        _table = table;
        _rows = new List<Row>(table.Rows);
    }

    public Query Select(params string[] columns)
    {
        foreach (var col in columns)
            _table.ValidateColumn(col);

        _rows = _rows.Select(r => r.Project(columns)).ToList();
        return this;
    }

    public Query Where(string column, string op, string value)
    {
        _table.ValidateColumn(column);
        _rows = _rows.Where(r => Evaluate(r[column], op, value)).ToList();
        return this;
    }

    public Query OrderBy(string column, string direction)
    {
        _table.ValidateColumn(column);
        var ascending = direction.Equals("asc", StringComparison.OrdinalIgnoreCase);

        _rows = _rows.OrderBy(r => r[column], new SmartComparer(ascending)).ToList();
        return this;
    }

    public Query Limit(int n)
    {
        _rows = _rows.Take(n).ToList();
        return this;
    }

    public int Count() => _rows.Count;

    public IReadOnlyList<Row> Rows => _rows;

    private static bool Evaluate(string cellValue, string op, string filterValue)
    {
        if (IsNumeric(cellValue) && IsNumeric(filterValue))
        {
            var cell = double.Parse(cellValue, CultureInfo.InvariantCulture);
            var filter = double.Parse(filterValue, CultureInfo.InvariantCulture);
            return op switch
            {
                "=" => cell == filter,
                "!=" => cell != filter,
                ">" => cell > filter,
                "<" => cell < filter,
                ">=" => cell >= filter,
                "<=" => cell <= filter,
                _ => false
            };
        }

        var cmp = string.Compare(cellValue, filterValue, StringComparison.Ordinal);
        return op switch
        {
            "=" => cmp == 0,
            "!=" => cmp != 0,
            ">" => cmp > 0,
            "<" => cmp < 0,
            ">=" => cmp >= 0,
            "<=" => cmp <= 0,
            _ => false
        };
    }

    private static bool IsNumeric(string value) =>
        double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out _);

    private class SmartComparer : IComparer<string>
    {
        private readonly int _direction;

        public SmartComparer(bool ascending)
        {
            _direction = ascending ? 1 : -1;
        }

        public int Compare(string? x, string? y)
        {
            if (x == null && y == null) return 0;
            if (x == null) return -1 * _direction;
            if (y == null) return 1 * _direction;

            if (IsNumeric(x) && IsNumeric(y))
            {
                var nx = double.Parse(x, CultureInfo.InvariantCulture);
                var ny = double.Parse(y, CultureInfo.InvariantCulture);
                return nx.CompareTo(ny) * _direction;
            }

            return string.Compare(x, y, StringComparison.Ordinal) * _direction;
        }
    }
}

File: csv-query/typescript/src/Query.ts
import type { CsvTable } from './CsvTable.js';
import type { Row } from './Row.js';

function isNumeric(value: string): boolean {
  return value.trim() !== '' && !isNaN(Number(value));
}

function evaluate(cellValue: string, op: string, filterValue: string): boolean {
  if (isNumeric(cellValue) && isNumeric(filterValue)) {
    const cell = Number(cellValue);
    const filter = Number(filterValue);
    switch (op) {
      case '=': return cell === filter;
      case '!=': return cell !== filter;
      case '>': return cell > filter;
      case '<': return cell < filter;
      case '>=': return cell >= filter;
      case '<=': return cell <= filter;
      default: return false;
    }
  }

  const cmp = cellValue < filterValue ? -1 : cellValue > filterValue ? 1 : 0;
  switch (op) {
    case '=': return cmp === 0;
    case '!=': return cmp !== 0;
    case '>': return cmp > 0;
    case '<': return cmp < 0;
    case '>=': return cmp >= 0;
    case '<=': return cmp <= 0;
    default: return false;
  }
}

export class Query {
  private readonly table: CsvTable;
  private currentRows: Row[];

  constructor(table: CsvTable) {
    this.table = table;
    this.currentRows = [...table.rows];
  }

  select(...columns: string[]): this {
    for (const col of columns) this.table.validateColumn(col);
    this.currentRows = this.currentRows.map(r => r.project(columns));
    return this;
  }

  where(column: string, op: string, value: string): this {
    this.table.validateColumn(column);
    this.currentRows = this.currentRows.filter(r => evaluate(r.get(column), op, value));
    return this;
  }

  orderBy(column: string, direction: string): this {
    this.table.validateColumn(column);
    const asc = direction.toLowerCase() === 'asc';
    this.currentRows.sort((a, b) => {
      const va = a.get(column);
      const vb = b.get(column);
      let cmp: number;
      if (isNumeric(va) && isNumeric(vb)) {
        cmp = Number(va) - Number(vb);
      } else {
        cmp = va < vb ? -1 : va > vb ? 1 : 0;
      }
      return asc ? cmp : -cmp;
    });
    return this;
  }

  limit(n: number): this {
    this.currentRows = this.currentRows.slice(0, n);
    return this;
  }

  count(): number {
    return this.currentRows.length;
  }

  get rows(): readonly Row[] {
    return this.currentRows;
  }
}
